### PR TITLE
Event Ring: multi-segment + dynamic ERSTSZ support (with tests)

### PR DESCRIPTION
**Pull request by ziyifu225**
This PR upgrades the Event Ring from single segment to multi-segment and adds support for dynamic ERST size changes at runtime (grow, shrink, overwrite entries). The ring now advances across segments, wraps correctly to segment 0, and flips the producer cycle bit only on full wraps. Dynamic changes take effect on the next segment boundary, keeping the ring usable while ERST is updated.

### Key changes

- Implement multi-segment traversal and wrap logic; re-read ERST on each segment transition.
- When not configured, the default value for ERSTSZ is 1.

### Tests added

- Multi-segment: single wrap, multiple wraps (cycle toggles).
- Ring-full detection (boundary case).
- Dynamic ERST: grow 1→3, shrink →1, and overwrite ERST entries;
- Defaulting behavior: ERSTSZ defaults to 1 when unset.

### Validation
Verified multi-segment behavior on a real USB device

When segment 0 is full, enqueue pointer advance to the next segment(segment 1). And cycle status isn't toggled.(cycle status = true)
image_009

When the ring is full, enqueue pointer wrap to segment 0 and cycle status is toggled.(cycle status = false)
image1

Successfully wraparound
image

> In my tests with a real USB device/guest, the Event Ring Segment Table Size (ERSTSZ) consistently ends up as 2. It appears to be programmed by the guest xHCI driver, not by our device code.
I haven’t found a reliable way to make the guest request >2 segments for validation. So far I’ve only exercised and verified the 2-segment path.

**Questions for reviewers**

1. Is “ERSTSZ = 2 in practice” expected for common guest kernels?
2. Do we have a recommended method to drive a larger ERSTSZ (e.g., kernel config/param, known guest, or a reproducible setup)?
3. For now, is validating with 2 segments considered sufficient?

### CODE REVIEW

**Review by lbeierlieb — DISMISSED** [review]
Very cool! I also verified passthrough with a real USB stick works and the event ring wraps around multiple times.

I left a few comments already, but my main request right now is to further split up the commits. Currently, they are not as "atomic" as they could be. Some pointers:
- the HCSPARAMS2 constant and associated XHCI MMIO read can be the last commit, basically enabling all of the previously extended functionality.
- if you modify/extend tests to cover new/changed functionality, please check the result of the old test first. If it still works, you can put the test changes into a follow-up commit. If the implementation changes break the tests, you should put the necessary test adaptations in the same commit (so that all tests pass on each commit). Also, if you just extend functionality (as here with multiple segments), you should double-check why tests start failing.

File: src/device/pci/rings.rs [hunk 1]
```diff
@@ -105,50 +100,65 @@ impl EventRing {
...
             .read(Request::new(erstba.wrapping_add(SIZE), RequestSize::Size4))
             as u32;
         self.cycle_state = true;
-        self.enqueue_pointer = self.segment_base;
-        self.trb_count = self.segment_size;
+
+        if self.erst_size == 0 {
+            self.set_erst_size(1);
+            warn!("ERSTSZ not set; defaulting to 1.");
+        }
```
[the remark this answers is not in the document]
Comment (ziyifu225) [reply src/device/pci/rings.rs]: You are right. the spec don't mention this. The device is now configured to crash on misconfiguration.

File: src/device/pci/rings.rs [hunk 2]
```diff
@@ -105,50 +100,65 @@ impl EventRing {
...
-    /// only segment, the trb_count to
+    /// Besides setting the base address of the Event Ring Segment Table, this
+    /// method initializes `enqueue_pointer` to the start of segment 0 and
+    /// sets `trb_count` from `ERST[0]`.
     ///
     /// # Parameters
     ///
-    /// - `erstba`: base address of the Event Ring Segment Table
+    /// - `erstba`: base address of the Event Ring Segment Table (ERST).
+    #[allow(clippy::cognitive_complexity)]
```
[the remark this answers is not in the document]
Comment (ziyifu225) [reply src/device/pci/rings.rs:122]: Done ✅

File: src/device/pci/rings.rs [hunk 3]
```diff
@@ -169,6 +181,11 @@ impl EventRing {
         self.dequeue_pointer
     }
 
+    /// Handle reads to the Event Ring Segment Table Size (ERSTSZ).
+    pub const fn read_erst_size(&self) -> u32 {
```
[the remark this answers is not in the document]
Comment (ziyifu225) [reply src/device/pci/rings.rs]: Done✅

File: src/device/pci/rings.rs [hunk 4]
```diff
@@ -67,6 +67,12 @@ pub struct EventRing {
     /// segment---because we only support one, we move back to the start of the
     /// same segment.
     trb_count: u32,
+    /// The index of the Event Ring segment currently being filled.
```
[the remark this answers is not in the document]
Comment (ziyifu225) [reply src/device/pci/rings.rs]: Done ✅

File: src/device/pci/xhci.rs [hunk 5]
```diff
@@ -482,7 +485,7 @@ impl PciDevice for Mutex<XhciController> {
             offset::CAPLENGTH => OP_BASE,
             offset::HCIVERSION => capability::HCIVERSION,
             offset::HCSPARAMS1 => capability::HCSPARAMS1,
-            offset::HCSPARAMS2 => 0, /* ERST Max size is a single segment */
+            offset::HCSPARAMS2 => capability::HCSPARAMS2, /* ERST Max size */
```
[the remark this answers is not in the document]
Comment (ziyifu225) [reply src/device/pci/xhci.rs]: Done ✅
